zkg

```diff
@@ -1469,7 +1469,16 @@ def cmd_upgrade(manager, args, config, configfile):
         for info, version, _ in outdated_packages:
             name = info.package.qualified_name()
 
-            if "test_command" not in info.metadata:
+            # Get info object for the next version as there may have been a
+            # test_command added during the upgrade.
+            next_info = manager.info(name, version=version, prefer_installed=False)
+            if next_info.invalid_reason:
+                print_error(
+                    f'error: invalid package "{name}": {next_info.invalid_reason}'
+                )
+                sys.exit(1)
+
+            if "test_command" not in next_info.metadata:
```

Comment:
This is slightly hacky since one always needs to remember that the `info` from `outdated_packages` might be outdated (which lead exactly to the bug you are fixing here). I wonder whether putting e.g., `[installed_version, next_version]` into `outdated_packages` would be harder to misuse. That would also help keeping that extra error handling code you added here in a single place.

Reply:
I've split the fix into a separate commit, but otherwise skipped changing anything around here. That doesn't mean I'm overly happy with that is, but any improvements would involve changing `outdated_packages` to actually contain namedtuples and I didn't want to go there.